Add additional events for controlling picker indicator on web

redstar504 opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details

I had to work through this a bit more to make the picker indicator behave consistently all platforms. As proposed, I added two new event handlers specific to the web version of the app that update the `isOpen` state, but they needed to be adapted a bit from what I suggested.

The original proposal worked to solve the reported issue, but it was causing the picker to remain highlighted on desktop web after selecting an item. To solve this, I switched from using `onClick` to `onMouseDown`. This solved it simply because of the order the event handlers are called in.

In addition, I included a new `onChange` event handler whose presence overrides the `onValueChange` handler for web. This was necessary because my original solution of setting `isOpen` to `false` in the `onValueChange` event was resulting in the  indicator changing back to the default setting whenever a new item was scrolled to in the iOS picker wheel.

### Fixed Issues
<!---
1. Please replace GH_LINK with a URL link to the GitHub issue this Pull Request is fixing.
2. Please replace PROPOSAL: GH_LINK_ISSUE(COMMENT) with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>
$ https://github.com/Expensify/App/issues/<number-of-the-issue(comment)>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/App/issues/15506
PROPOSAL: https://github.com/Expensify/App/issues/15506#issuecomment-1451139715


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. Login to the app and navigate to any page containing a Picker component.
2. In this issue, we used the `Settings > Workspace > Reimburse Expenses` picker for selecting mileage units.
3. Open the picker and confirm that the green indicator beneath the picker is bright green while the picker is open.
4. Select an item.  Confirm that after selecting an item the indicator returns to it's default dark green color.

- [X] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
5. Upload an image via copy paste
6. Verify a modal appears displaying a preview of that image
--->

1. Login to the staging environment and navigate to any page containing a Picker component.
2. In this issue, we used the `Settings > Workspace > Reimburse Expenses` picker for selecting mileage units.
3. Open the picker and confirm that the green indicator beneath the picker is bright green while the picker is open.
4. Select an item.  Confirm that after selecting an item the indicator returns to it's default dark green color.

- [X] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android / native
    - [X] Android / Chrome
    - [X] iOS / native
    - [X] iOS / Safari
    - [X] MacOS / Chrome / Safari
    - [X] MacOS / Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

#### Safari

https://user-images.githubusercontent.com/1311325/222996647-5df944a0-fcb7-4a4a-8d27-025e10f745e5.mp4

#### Chrome

Having an issue with OBS not displaying `<select>` elements during a windowed screen recording (seems like `select` elements are considered a separate window), but still demonstrates the functionality:

https://user-images.githubusercontent.com/1311325/223631008-f2da367a-34e0-45b7-b7b3-fcf28453efe6.mp4

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/1311325/223631057-a5b5df0b-87d9-4d1a-af0e-742be584be31.mp4

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/1311325/222996270-192ab927-379d-46f2-bdf0-c379928f2032.mp4

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/1311325/223631123-6fec3bae-890c-4142-96fe-8ad0eb67c042.mp4

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/1311325/223631143-ce7c083a-e397-49d3-a465-4a5b17b71c3c.mp4

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/1311325/223631164-b97b1458-d075-463b-be23-01dc04c2722a.mp4

</details>

redstar504 committed:
Add additional events for controlling picker indicator on web

marcaaron commented on src/components/Picker.js:
```diff
@@ -193,6 +193,21 @@ class Picker extends PureComponent {
                             },
+
+                            // The following 2 handlers are specific to web (onChange overrides onValueChange on web)
+                            // They are used to make the picker indicator behave the same way as it does on native
```
>To solve this, I switched from using onClick to onMouseDown. This solved it simply because of the order the event handlers are called in.

Can we also briefly mention something about this in this code comment? Feels possible someone might make the mistake of switching back to an `onClick` behavior without this.

marcaaron replied:
Also, how does `onChange` override `onValueChange` on web?

redstar504 replied:
I added some comments regarding this in the `index.js` file of my latest commit.

Let me know if you have any more questions.

marcaaron commented on src/components/Picker.js:
```diff
@@ -193,6 +193,21 @@ class Picker extends PureComponent {
+                                const index = e.target.selectedIndex;
+                                const value = e.target.options[index].value;
+                                this.onInputChange(value, index);
+                                this.setState({isOpen: false});
```
It feels like we are sneaking in some platform dependent code here, but it's not clear which is the web or native implementation. Normally, [how we deal with that](https://github.com/expensify/app#platform-specific-file-extensions) is create a "base" component and then an `index.js` and `index.native.js` components. In a file like `Picker.js` we'd then just import the Picker but the platform dependent changes would be contained in the respective versions of the picker. Would this approach work here?

marcaaron commented:
@thesahindia I don't feel super strongly about solving that case in this ticket. We didn't spot it until now and it's not clear what the solution would be - so I say we continue here and then file it as a separate bug. Unless @redstar504 has some idea on how to tackle it and is willing to look into it.

redstar504 commented:
@marcaaron @thesahindia Thanks for the input guys. I have been exploring ways to split up the platform-specific code as requested. It appears that our goal is to share the Picker component across all platforms while providing a means to specify any platform-specific event handlers through their respective index files.

Would it make sense to refactor `Picker.js` into `Picker/BasePicker.js`, containing all of the platform independent handlers, and then import `BasePicker` into the platform specific index files? We could pass any platform specific event handlers to `BasePicker` as props to enable customization for each platform.

redstar504 commented:
My latest commit includes platform specific event handling for the Picker component. The implementation involves wrapping the refactored BasePicker with platform specific entrypoint files, and wrapping the entrypoints with a higher order component that supplies default events for both platforms. Web specific events are encapsulated in the index.js file. Native events are supplied as defaults through the HOC since they supplement the web-based events.

Testing on all platforms was successful, and I have included new recordings in my checklist.

@marcaaron @thesahindia

marcaaron commented on src/components/withPickerEvents.js:
```diff
@@ -0,0 +1,120 @@
+import {pickerPropTypes, pickerDefaultProps} from './Picker/pickerPropTypes';
+
+export default function (WrappedComponent) {
+    class WithSharedEvents extends Component {
```
I haven't delved deeply into the reasons for needing it, but is an HOC necessary? Feels like the "WrappedComponent" is always going to be the `Picker`.

Additionally, I'm having trouble following some of the conventions used here e.g. the `callbackConsumer()` (stuck out to me as unusual for our codebase and my instincts are telling me it could confuse people).

I'd have to do a deeper dive to come up with more concrete suggestions, but just wanted to toss that observation out there that maybe we could simplify this.

redstar504 replied:
The HOC wraps the Picker in both entrypoints (index.js and index.native.js), so that we can share all of the code contained in it between them. This also gives us a way to manipulate the events passed down to BasePicker using  an entrypoint file. I originally had placed the HOC in the Picker directory because it would only ever be used for the Picker, but I noticed all of the others were contained in the base `src` directory so I moved it there.

An HOC would seem like overkill but I am kind of stuck between a dead simple solution (the one I originally provided), and an over-engineered solution.  I have been having trouble finding something in the middle ground that won't introduce a lot of code duplication.  Another option might be class inheritance.

We basically need a way to share all of the event handlers between both entrypoints but give the web entry point a way to extend/override them before passing them down to BasePicker.  If you have any alternative ideas for that please let me know. If we stick with the HOC I can definitely rename `callbackConsumer` back to `onInputChange`.

marcaaron replied:
Ok I see the issue now - sorry for the distraction here. I think where this is at is a bit hard to reason about for not much gain.

Here's what I'm thinking as an alternative...

1. Start by reverting what you have here back to https://github.com/Expensify/App/pull/15673/commits/15fb9c7db99ac31f24b6b78926115a1a4cca5e8b
1. Create some prop like `shouldUseAdditionalPickerEvents` on the Picker
1. Override the methods that way inside the Picker

It might seem vain, but we can get around the cross platform file extension convention this way and it will allow web/native to switch on/off the functionality we need without specifically referencing the platform anywhere.

redstar504 replied:
@marcaaron Thank you for the advice.  I just need a bit more clarification:

> It might seem vain, but we can get around the cross platform file extension convention this way and it will allow web/native to switch on/off the functionality we need without specifically referencing the platform anywhere.

Are you suggesting not having platform-specific entry points in this case?  I am just curious what is going to toggle `shouldUseAdditionalPickerEvents` if we don't, and if we didn't, where do we define the web-specific events?

redstar504 replied:
@marcaaron Ok I will give that a shot.  Since it's just a conditional in the entry points, is this implying that we keep the web events defined in the Picker itself?

marcaaron replied:
>implying that we keep the web events defined in the Picker itself

That was my meaning yea.

But I thought of something that is slightly better and might result in fewer changes while avoiding a reference to the web specific `e.target`...

- Create `getAdditionalPickerEvents()`
    - have native return `{}`
    - have web return something like

```js
// getAdditionalPickerEvents/index.js
function getAdditionalPickerEvents(onMouseDown, onChange) {
    return {
        onMouseDown,
        onChange: (e) = {
            if (e.target.selectedIndex === undefined) {
                return;
            }
            const index = e.target.selectedIndex;
            const value = e.target.options[index].value;
            onChange(value, index);
        },
    };
}
```

Then pass them like:
```js
pickerProps={{
    ...getAdditionalPickerEvents(this.setIsOpen, this.updateAndClose),
```
  
This feels slightly better to me as it manages the web event outside of the cross platform Picker code. Sorry for all the back and forth. Our style guide sometimes paints us into a corner 😄

redstar504 replied:
@marcaaron I like the looks of that.  I have been trying to find a way to pass down the events from the web entry, but also to be able to pass a function back that can enable/disable the highlight without having to reference them from inside the shared Picker.  This looks like it could work.

Also, no worries. I am happy to work towards something we can all be satisfied with.

redstar504 committed:
Add web-platform specific Picker events using HOF

redstar504 commented:
That concept worked perfectly @marcaaron, nice job.

The way I implemented it is by passing the `additionalPickerEvents` containing the web events through props from the `index.js` entrypoint.  In the Picker, I declared `additionalPickerEvents` as a `defaultProp` with an empty function, so we don’t have to pass anything from the native entrypoint.

I refactored all of the `setState({isOpen: bool})` calls into setters like in your example.  I did not introduce an `updateAndClose()` function, as the native implementation does not handle both operations in a single event, so it would have only been used for web.

Again, checks out on all platforms.

marcaaron commented on src/components/Picker/Picker.js:
```diff
@@ -151,6 +155,18 @@ class Picker extends PureComponent {
+        this.setState({
+            isOpen: false,
+        });
+    }
```
Can we bind these in the constructor?

marcaaron commented on src/components/Picker/Picker.js:
```diff
@@ -184,15 +200,22 @@ class Picker extends PureComponent {
                         fixAndroidTouchableBug
-                        onOpen={() => this.setState({isOpen: true})}
-                        onClose={() => this.setState({isOpen: false})}
+                        onOpen={() => this.setIsOpen()}
+                        onClose={() => this.setIsClosed()}
```
```suggestion
                        onOpen={this.setIsOpen}
                        onClose={this.setIsClosed}
```

marcaaron commented on src/components/Picker/Picker.js:
```diff
@@ -184,15 +200,22 @@ class Picker extends PureComponent {
                         textInputProps={{allowFontScaling: false}}
                         pickerProps={{
-                            onFocus: () => this.setState({isOpen: true}),
+                            onFocus: () => this.setIsOpen(),
```
```suggestion
                            onFocus: this.setIsOpen,
```

marcaaron commented on src/components/Picker/Picker.js:
```diff
@@ -184,15 +200,22 @@ class Picker extends PureComponent {
                                 this.props.onBlur();
                             },
+                            ...this.props.additionalPickerEvents(
+                                () => this.setIsOpen(),
```
```suggestion
                                this.setIsOpen,
```

marcaaron reviewed (changes requested):
Looks great to me. Just a couple of places where I think we could use a bound function instead of an anonymous arrow function.

redstar504 committed:
Use bound functions for setting opened state

redstar504 commented:
@marcaaron Requested changes pushed!

marcaaron reviewed (dismissed):
Didn't test but code LGTM so approving. All yours @thesahindia.

redstar504 committed:
Clarify highlighted state, reusable props

redstar504 commented on src/components/Picker/Picker.js:
```diff
@@ -151,6 +157,18 @@ class Picker extends PureComponent {
+        this.setState({
+            isOpen: false,
+        });
+    }
```
You are right, I think we should clarify what that state variable actually controls.  Since it is only responsible for enabling and disabling the highlight, we can make it more understandable by using an `isHighlighted` state variable, and calling `enableHighlight` and `disableHighlight` to control it.

redstar504 commented on src/components/Picker/Picker.js:
```diff
@@ -3,13 +3,13 @@ import React, {PureComponent} from 'react';
+import themeColors from '../../styles/themes/default';
+import {ScrollContext} from '../ScrollViewWithContext';
 
 const propTypes = {
```
Since no props are specifically referenced in the platform-specific wrappers, it might be unnecessary to perform validation there before passing them to the BasePicker. However, if this is an established pattern in the codebase, it would be a good idea to follow that with our changes as well.

marcaaron replied:
I think I don't feel too passionate about this one. Or at least I can't think of a situation where it would help because the props are just getting passed straight to the BasePicker. If there was some prop validation warning we would see it either way.

redstar504 commented:
Thanks for the review @thesahindia.  I have added your requested changes to the latest commit.

redstar504 commented:
@thesahindia

The error is due to the prop validation we've added in the entrypoints.  Adding `propTypes` directly to the component created by `forwardRef` does not validate as expected. It seems that we would have to wrap the BasePicker, do the prop validation and then wrap it again using `forwardRef`.  This would be unnecessary, since all of the props are already being validated in the `BasePicker`. There is no need to validate them in the entrypoints as well.

For example, this works but does not provide any benefit:
```JavaScript
// index.js
import {defaultProps, propTypes} from './pickerPropTypes';

// eslint-disable-next-line react/jsx-props-no-spreading
const Picker = props => <BasePicker {...props} additionalPickerEvents={additionalPickerEvents} />;

Picker.propTypes = propTypes;
Picker.defaultProps = defaultProps;

// eslint-disable-next-line react/jsx-props-no-spreading
export default forwardRef((props, ref) => <Picker {...props} innerRef={ref} />);
```

Using shared `propTypes` looks to be common when there are multiple entrypoints that do not wrap an underlying component.  It's purpose would be to give them a common interface to conform to. 

I suggest that we rollback this change as it is not necessary for our scenario.

redstar504 committed:
Revert shared propTypes

redstar504 commented:
@thesahindia I reverted shared prop types due to the reasons described in my last comment.  Props are still validated as expected through the BasePicker component, and there are no console errors.  So we should be good.

thesahindia reviewed (approved):
~Looks good and tests well!~

~@marcaaron, I have filled the checklist. It's all yours!~

thesahindia reviewed (changes requested):
I take back what I said. It didn't test well. I was checking other pages to confirm that all the functionalities are working as expected. It is not working correctly. Not getting the error message for the picker. @redstar504, can you look into it?

I am on a slow internet so I couldn't upload the video. You can try reproducing it at "Add debit card page" or any other page where we are validating.

redstar504 committed:
Pass ref prop properly to base component

redstar504 commented:
@thesahindia Great job catching that.  We needed to pass the `ref` all the way down to the base component by that name to continue forwarding it, and only then it's passed to the RNPickerSelect as `innerRef`.  It should be fixed now.

thesahindia reviewed (approved):
Thanks! It works well.

cc: @marcaaron

marcaaron reviewed (approved):
Nice job here overall and LGTM. Thanks @thesahindia for the comments and testing 🙇

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/4425479651) to staging by https://github.com/marcaaron in version: 1.2.85-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/4436166249) to production by https://github.com/Julesssss in version: 1.2.85-1 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
